### PR TITLE
[PropertyWrappers] Look for constructors with default args when finding default init 

### PR DESCRIPTION
**Pull request by theblixguy**
At the moment, `findDefaultInit()` only looks for `init()` and ignores any constructors that have default args. This is a [bug](https://twitter.com/dgregor79/status/1172533772291473409) and it prevents the following code from working:

```swift
@propertyWrapper
struct Foo {
  let name: String

  init(name: String = "hello world") {
    self.name = name
  }

  var wrappedValue: Int {
    get { return 0 }
  }
}

struct Bar {
  @Foo var property: Int // error: property not initialized
  init() {}
}
```

When finding default initializers, also search for constructors with default args.

Resolves [SR-11477](https://bugs.swift.org/browse/SR-11477).

### CODE REVIEW

**Comment by theblixguy** [comment]
I also noticed we didn't have any tests for multiple default inits, so I have now added some typechecker tests as well. I am going to run a quick test now while I wait for @DougGregor's review. 

@swift-ci please smoke test

**Comment by theblixguy** [comment]
@swift-ci please smoke test

**Comment by theblixguy** [comment]
@swift-ci please test source compatibility

File: lib/Sema/TypeCheckPropertyWrapper.cpp
Comment (DougGregor) [inline lib/Sema/TypeCheckPropertyWrapper.cpp]: Why skip the synthesized default initializer? It's reasonable to consider calling it, so long as it's internal.

File: lib/Sema/TypeCheckPropertyWrapper.cpp
Comment (DougGregor) [inline lib/Sema/TypeCheckPropertyWrapper.cpp]: I don't think we need to check `hasParams`? `allParametersHaveDefaultArg` is trivially `true` if there are 0 parameters.

There might be a failure case here where the initializer has generic parameters that won't get inferred if you don't pass a parameter, which means it won't be able to be called. For example `init<T: ExpressibleByIntegerLiteral>(_ value: T = 0)`

**Review by DougGregor — CHANGES_REQUESTED** [review]

File: lib/Sema/TypeCheckPropertyWrapper.cpp
Comment (theblixguy) [reply lib/Sema/TypeCheckPropertyWrapper.cpp]: We might have a init() and a synthesised init(wrappedValue:), which now triggers the “multiple default value initialiser” unless we ignore the synthesised init.

File: lib/Sema/TypeCheckPropertyWrapper.cpp
Comment (theblixguy) [reply lib/Sema/TypeCheckPropertyWrapper.cpp]: Ah, you’re right. I’ll remove the check. I suppose we can skip an init() where it has generic parameters (and default arguments).

File: lib/Sema/TypeCheckPropertyWrapper.cpp
Comment (DougGregor) [reply lib/Sema/TypeCheckPropertyWrapper.cpp]: Hmm. We could use `TypeChecker::compareDeclarations` to pick the best default-value initializer. That seems more principled than ignoring implicit inits.

File: lib/Sema/TypeCheckPropertyWrapper.cpp
Comment (DougGregor) [reply lib/Sema/TypeCheckPropertyWrapper.cpp]: Thanks!

File: lib/Sema/TypeCheckPropertyWrapper.cpp
Comment (theblixguy) [reply lib/Sema/TypeCheckPropertyWrapper.cpp]: Sounds like a good idea! I have done that now and it seems to work as expected.

Comment (xwu) [reply lib/Sema/TypeCheckPropertyWrapper.cpp]: What makes something `init1` versus `init3`? Source order? That would be brittle.

Seems to me that in the rare circumstance of many equally acceptable initializers, it would be reasonable to have the user explicitly clarify.

Comment (theblixguy) [reply lib/Sema/TypeCheckPropertyWrapper.cpp]: The initializers are compared using `Typechecker::compareDeclarations`, which returns `Comparison::Unordered` if two decls are “equal” (they’re not better or worse than each other). See CSRanking.cpp to understand how it works.

I could bail out if we ever end up in that situation, instead of being greedy and using whatever “best” initializer we’ve found so far.

**Comment by theblixguy** [comment]
@DougGregor Do you have any further feedback on this?

**Comment by DougGregor** [comment]
Sorry, I caused a conflict when I merged https://github.com/apple/swift/pull/27322

**Comment by theblixguy** [comment]
@DougGregor No problem, I have rebased now 👍 Can you take another look at this?

File: test/decl/var/property_wrappers.swift
Comment (natecook1000) [inline test/decl/var/property_wrappers.swift]: Do we have a test that verifies which initializer is picked by writing `@SR_11477_W1 var foo: Int`?

File: lib/Sema/TypeCheckPropertyWrapper.cpp
Comment (DougGregor) [inline lib/Sema/TypeCheckPropertyWrapper.cpp]: I know I suggested this, but having gone down the path of https://github.com/apple/swift/pull/27322, I've convinced myself that's the better approach: rather than having another "pick the best declaration" implementation here, check whether any of the initializers fit the form... and then record the presence of such an initializer rather than which one was picked. Let ambiguity resolution and diagnostics be the responsibility of the expression type checker. I bet you could even generalize `findInitialValueInit` to cover the default-init case, and end up with a net deletion of code in the compiler. What do you think?

File: test/decl/var/property_wrappers.swift
Comment (theblixguy) [reply test/decl/var/property_wrappers.swift]: No, but I have added one now!

File: lib/Sema/TypeCheckPropertyWrapper.cpp
Comment (theblixguy) [reply lib/Sema/TypeCheckPropertyWrapper.cpp]: I had the same thought yesterday... it sounds like a good idea! I think `findInitialValueInit` can already handle this because it requires all parameters to be either (1) default or (2) have the given argument label and we just care about (1).

**Comment by theblixguy** [comment]
@DougGregor I have combined both methods into one as per your suggestion. Does it look good to you?

File: lib/Sema/TypeCheckPropertyWrapper.cpp
Comment (DougGregor) [inline lib/Sema/TypeCheckPropertyWrapper.cpp]: I think you also want "&& !argumentLabel.empty()` here, otherwise with the default-init version it'll match an unlabeled parameter.

File: lib/Sema/TypeCheckPropertyWrapper.cpp
Comment (DougGregor) [inline lib/Sema/TypeCheckPropertyWrapper.cpp]: This `&=` doesn't do anything? We start at `allArgsDefault = true`, so this is a no-op.

File: lib/Sema/TypeCheckPropertyWrapper.cpp
Comment (DougGregor) [inline lib/Sema/TypeCheckPropertyWrapper.cpp]: The approach of setting `argumentParam = nullptr` in my PR was a little unclear, but when we have to also set `allArgsDefault = false` and separately check that for default initializers, it gets hard to reason about. What if we have an `bool extraneousParam = false` that gets set `tru` on the failure path, and check that below. If there's an extraneous parameter, regardless of which kind of initializer we're looking for, skip this initializer.

**Review by DougGregor — COMMENTED** [review]
This is greatly improved! A few minor comments about the control flow in the unified lookup, but this will be ready to go in. Thank you!

**Comment by theblixguy** [comment]
@DougGregor Thanks for the review! I have addressed your comments :)

**Comment by theblixguy** [comment]
ping @DougGregor, can you approve this if you're happy? thank you!

**Comment by theblixguy** [comment]
A conflict creeped in - fixed it! @DougGregor just waiting for your approval :-)

**Review by DougGregor — APPROVED** [review]
Thank you! Looks fantastic.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by theblixguy** [comment]
@swift-ci please smoke test

**Comment by theblixguy** [comment]
@swift-ci please smoke test

**Comment by theblixguy** [comment]
@swift-ci please test source compatibility

**Comment by theblixguy** [comment]
Debug source compat failure is an unrelated UPASS.

**Pull request merged**
